execution/scheduler_ext_test.go

```diff
@@ -1045,96 +1044,97 @@ func TestDNSResolver(t *testing.T) {
+			}},
+			expNewResolutions: 2,
+		},
+		"3s": {
+			opts: lib.Options{DNS: types.DNSConfig{
+				TTL:    null.StringFrom("3s"),
+				Select: types.NullDNSSelect{DNSSelect: types.DNSfirst, Valid: true},
+				Policy: types.NullDNSPolicy{DNSPolicy: types.DNSpreferIPv4, Valid: false},
+			}},
+			expNewResolutions: 1,
+		},
+	}
 
-				mr := mockresolver.New(nil, net.LookupIP)
-				runner.ActualResolver = mr.LookupIPAll
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			logger := logrus.New()
+			logger.SetOutput(io.Discard)
+			logHook := testutils.NewLogHook(logrus.WarnLevel)
+			logger.AddHook(logHook)
 
-				ctx, cancel, execScheduler, samples := newTestScheduler(t, runner, logger, tc.opts)
-				defer cancel()
+			registry := metrics.NewRegistry()
+			builtinMetrics := metrics.RegisterBuiltinMetrics(registry)
+			runner, err := js.New(
+				&lib.TestPreInitState{
+					Logger:         logger,
+					BuiltinMetrics: builtinMetrics,
+					Registry:       registry,
+				},
+				&loader.SourceData{
+					URL: &url.URL{Path: "/script.js"}, Data: []byte(script),
+				}, nil)
+			require.NoError(t, err)
 
-				mr.Set("myhost", sr("HTTPBIN_IP"))
-				time.AfterFunc(1700*time.Millisecond, func() {
+			mr := mockresolver.New(map[string][]net.IP{"myhost": {net.ParseIP(sr("HTTPBIN_IP"))}})
+			var newResolutions uint32
+			resolveHook := func(host string, result []net.IP) {
+				require.Len(t, result, 1)
+				if result[0].String() == "127.0.0.1" {
 					mr.Set("myhost", "127.0.0.254")
-				})
-				defer mr.Unset("myhost")
+				} else {
+					atomic.AddUint32(&newResolutions, 1)
+				}
+			}
+			mr.ResolveHook = resolveHook
+			runner.ActualResolver = mr.LookupIPAll
 
-				errCh := make(chan error, 1)
-				go func() { errCh <- execScheduler.Run(ctx, ctx, samples) }()
+			ctx, cancel, execScheduler, samples := newTestScheduler(t, runner, logger, tc.opts)
+			defer cancel()
 
-				select {
-				case err := <-errCh:
-					require.NoError(t, err)
-					entries := logHook.Drain()
-					require.Len(t, entries, tc.expLogEntries)
-					for _, entry := range entries {
-						require.IsType(t, &url.Error{}, entry.Data["error"])
-						err, ok := entry.Data["error"].(*url.Error)
-						assert.True(t, ok)
-						errMsg := err.Error()
-						contains := strings.Contains(errMsg, "connection refused") ||
-							strings.Contains(errMsg, "request timeout") ||
-							strings.Contains(errMsg, "dial: i/o timeout")
-						assert.Truef(t, contains, "expected error to contain one of the list of messages")
-					}
-				case <-time.After(10 * time.Second):
-					t.Fatal("timed out")
-				}
-			})
-		}
-	})
+			errCh := make(chan error, 1)
+			go func() { errCh <- execScheduler.Run(ctx, ctx, samples) }()
+
+			select {
+			case err := <-errCh:
+				require.NoError(t, err)
+				assert.Equal(t, tc.expNewResolutions, atomic.LoadUint32(&newResolutions))
+			case <-time.After(10 * time.Second):
```

Comment:
```suggestion
			case <-time.After(2 * time.Second):
```
nit: I think at this point we can cut the waiting time in case we will have a failure.